### PR TITLE
[AMORO-1966] Add PostgreSQL as system database

### PR DESCRIPTION
**Pull request by XBaith**
## Why are the changes needed?

Implement #1966 

## Brief change log

-

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java [hunk 1]
```diff
@@ -17,7 +17,7 @@ public class Object2ByteArrayConvert<T> implements TypeHandler<T> {
   @Override
   public void setParameter(PreparedStatement ps, int i, T parameter, JdbcType jdbcType) throws SQLException {
     if (parameter == null) {
-      ps.setNull(i, Types.BLOB);
+      ps.setNull(i, Types.BINARY);
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java:20]: What is the reason for this modification?

Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java:20]: Why should way make this change?

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java:20]: Postgres require `BINARY` for `bytea` column type

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java:20]:
```
### The error may exist in com/netease/arctic/server/persistence/mapper/OptimizingMapper.java (best guess)
### The error may involve com.netease.arctic.server.persistence.mapper.OptimizingMapper.insertTaskRuntimes-Inline
### The error occurred while setting parameters
### SQL: INSERT INTO task_runtime (process_id, task_id, retry_num, table_id, partition_data, start_time, end_time, status, fail_reason, optimizer_token, thread_id, rewrite_output, metrics_summary, properties) VALUES     (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)  ,  (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)  ,  (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)  ,  (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)
### Cause: org.postgresql.util.PSQLException: ERROR: column "rewrite_output" is of type bytea but expression is of type oid
  Hint: You will need to rewrite or cast the expression.
  Position: 259
	at org.apache.ibatis.exceptions.ExceptionFactory.wrapException(ExceptionFactory.java:30) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:199) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.insert(DefaultSqlSession.java:184) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:62) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.binding.MapperProxy$PlainMethodInvoker.invoke(MapperProxy.java:152) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:85) ~[mybatis-3.5.6.jar:3.5.6]
	at com.sun.proxy.$Proxy49.insertTaskRuntimes(Unknown Source) ~[?:?]
	at com.netease.arctic.server.optimizing.OptimizingQueue$TableOptimizingProcess.lambda$null$6(OptimizingQueue.java:571) ~[amoro-ams-server-0.5.0.jar:?]
	at com.netease.arctic.server.persistence.PersistentBase.doAs(PersistentBase.java:48) ~[amoro-ams-server-0.5.0.jar:?]
	... 26 more 
```

**Review by wangtaohz — COMMENTED** [review]
Thanks a lot for you contribution! 👍

I think we should add some config demo of PostgreSQL to `dist/src/main/arctic-bin/conf/config.yaml`.

Also, we can add some docs to `docs/admin-guides/deployment.md` about using PostgreSQL as system database.

File: ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java (hunk 1 above)
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java:20]: I think not, I test it locally and work fine

Comment (shidayang) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java:20]: ok

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java]: Let me check db type before set `NULL` value

**Comment by XBaith** [comment]
> I think we should add some config demo of PostgreSQL to `dist/src/main/arctic-bin/conf/config.yaml`.
> 
> Also, we can add some docs to `docs/admin-guides/deployment.md` about using PostgreSQL as system database.

Sure, I will add more details to related documents

**Comment by zhoujinsong** [comment]
Can you change the prefix of PR title to `[AMORO-1966]`?

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java [hunk 2]
```diff
@@ -127,6 +132,9 @@ private void createTablesIfNeed(Configurations config) {
         query = String.format(
             "SELECT 1 FROM information_schema.tables WHERE table_schema = '%s' AND table_name = '%s'",
             connection.getCatalog(), "CATALOG_METADATA");
+      } else if (ArcticManagementConf.DB_TYPE_POSTGRES.equals(dbTypeConfig)) {
+        query = "SELECT 1 FROM information_schema.tables WHERE table_schema = 'public' AND table_name = " +
+            "'catalog_metadata'";
```
Comment (baiyangtx) [inline ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:137]: Is the database information missing? And is the schema fixed as public?

File: ams/server/src/main/resources/postgres/ams-postgres-init.sql [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+);
+COMMENT ON TABLE optimizer IS 'Optimizer metadata';
+COMMENT ON COLUMN optimizer.resource_id IS 'Optimizer instance ID';
+COMMENT ON COLUMN optimizer.group_name IS 'Group/queue name';
+COMMENT ON COLUMN optimizer.container_name IS 'Container name';
+COMMENT ON COLUMN optimizer.start_time IS 'Optimizer start time';
+COMMENT ON COLUMN optimizer.touch_time IS 'Update time';
+COMMENT ON COLUMN optimizer.thread_count IS 'Total number of all CPU resources';
+COMMENT ON COLUMN optimizer.total_memory IS 'Optimizer memory usage';
+COMMENT ON COLUMN optimizer.properties IS 'optimizer state info, contains like yarn application id and flink job id';
```
Comment (baiyangtx) [inline ams/server/src/main/resources/postgres/ams-postgres-init.sql:55]: There is an Index on `group_name` for mysql scripts.

Is the index on "group_name" necessary?

File: ams/server/src/main/resources/postgres/ams-postgres-init.sql [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+COMMENT ON COLUMN resource.properties IS 'Optimizer instance properties';
+
+CREATE TABLE resource_group
+(
+    group_name varchar(50) NOT NULL,
+    container_name varchar(100) DEFAULT NULL,
+    properties TEXT,
+    PRIMARY KEY (group_name)
+);
+CREATE INDEX resource_group_index ON resource_group (group_name);
```
Comment (baiyangtx) [inline ams/server/src/main/resources/postgres/ams-postgres-init.sql:85]: You have an Index on Primary Key.

File: ams/server/src/main/resources/postgres/ams-postgres-init.sql [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    summary TEXT,
+    from_sequence TEXT,
+    to_sequence TEXT,
+    PRIMARY KEY (process_id)
+);
+COMMENT ON TABLE table_optimizing_process IS 'History of optimizing after each commit';
+COMMENT ON COLUMN table_optimizing_process.process_id IS 'Optimizing procedure UUID';
+COMMENT ON COLUMN table_optimizing_process.table_id IS 'Table ID';
+COMMENT ON COLUMN table_optimizing_process.catalog_name IS 'Catalog name';
+COMMENT ON COLUMN table_optimizing_process.db_name IS 'Database name';
```
Comment (baiyangtx) [inline ams/server/src/main/resources/postgres/ams-postgres-init.sql:222]: There is an Index on (`table_id`, `plan_time`) in MySQL script.

Please help to check if this index is necessary.

File: ams/server/src/main/resources/postgres/ams-postgres-init.sql [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    status VARCHAR(16),
+    fail_reason VARCHAR(4096),
+    optimizer_token VARCHAR(50),
+    thread_id INT,
+    rewrite_output BYTEA,
+    metrics_summary TEXT,
+    properties TEXT,
+    PRIMARY KEY (process_id, task_id)
+);
+COMMENT ON TABLE task_runtime IS 'Optimize task basic information';
```
Comment (baiyangtx) [inline ams/server/src/main/resources/postgres/ams-postgres-init.sql:256]: There is an Index on (`table_id`, `process_id`) in MySQL script.

File: ams/server/src/main/resources/postgres/ams-postgres-init.sql [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    task_id INT NOT NULL,
+    retry_num INT DEFAULT 0,
+    table_id BIGINT NOT NULL,
+    start_time TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    end_time TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    fail_reason VARCHAR(4096),
+    PRIMARY KEY (process_id, task_id, retry_num)
+);
+COMMENT ON TABLE optimizing_task_quota IS 'Optimize task basic information';
+COMMENT ON COLUMN optimizing_task_quota.process_id IS 'Optimizing procedure UUID';
```
Comment (baiyangtx) [inline ams/server/src/main/resources/postgres/ams-postgres-init.sql:286]: There is an Index on (`table_id`) in MySQL script.

File: ams/server/src/main/resources/postgres/ams-postgres-init.sql [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+(
+    blocker_id BIGSERIAL PRIMARY KEY,
+    catalog_name VARCHAR(64) NOT NULL,
+    db_name VARCHAR(128) NOT NULL,
+    table_name VARCHAR(128) NOT NULL,
+    operations VARCHAR(128) NOT NULL,
+    create_time TIMESTAMP,
+    expiration_time TIMESTAMP,
+    properties TEXT
+);
```
Comment (baiyangtx) [inline ams/server/src/main/resources/postgres/ams-postgres-init.sql:333]: There is an index on (`catalog_name`,`db_name`,`table_name`)

**Review by baiyangtx — CHANGES_REQUESTED** [review]
I think you forgot to COPY the newly added SQL scripts in dis/assemblies/bin.xml.

File: ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java (hunk 1 above)
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java]: In JDBC, Types.BINARY and Types.BLOB are constants used to represent binary data types. Their meanings differ slightly in MySQL and PostgreSQL, as follows:

In MySQL:

Types.BINARY: represents the BINARY type in MySQL, which is used to store fixed-length binary data.
Types.BLOB: represents the BLOB type in MySQL, which is used to store variable-length binary data.

When using MySQL, the ResultSet.getBytes() method can be used to retrieve data of BINARY or BLOB type. For BINARY type, the length of the returned byte array is equal to the length specified when the field was defined. For BLOB type, the length of the returned byte array is equal to the actual length of the data stored.

In PostgreSQL:

Types.BINARY: represents the BYTEA type in PostgreSQL, which is used to store variable-length binary data.
Types.BLOB: has no corresponding type in PostgreSQL, so the Types.BLOB constant should not be used when working with PostgreSQL.

When using PostgreSQL, the ResultSet.getBytes() method can be used to retrieve data of BYTEA type. The length of the returned byte array is equal to the actual length of the data stored.

In summary, the meanings of Types.BINARY and Types.BLOB differ slightly in MySQL and PostgreSQL, and the appropriate choice should be made based on the specific database type and use case. It is important to distinguish and use them correctly when working with these databases.


ChatGPT gave me the above answer.

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java (hunk 2 above)
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:137]: I think database informantion has been set in jdbc url

**Review by shidayang — APPROVED** [review]

File: ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java (hunk 2 above)
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:137]: So remove condition: `table_schema = 'public'` is better solution?

Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:137]: Is there someway to get the current table_schema ?

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:137]: Yes, execute SQL `SELECT current_schema()` will retuen the value

Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java:137]: So we can change the sql to `WHERE table_schema = current_schema() AND table_name = 'catalog_metadata' `

Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/persistence/SqlSessionFactoryProvider.java]: Fixed

**Review by baiyangtx — APPROVED** [review]
LGTM, thanks a lot for your contribution.
